Add documentation for a use case: "How To Pass a Physical NIC Into a Kata Container"

Yuval-Ai opened the pull request:
As discussed in [issue #1876](https://github.com/kata-containers/runtime/issues/1876).
And in [this comment](https://github.com/kata-containers/runtime/issues/1876#issuecomment-603575507) by @amshinde

**Note:** The document is [incomplete ](https://github.com/kata-containers/runtime/issues/1876#issuecomment-554744743)

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+# How To Pass a Physical NIC Into a Kata Container
+![](https://github.com/Yuval-Ai/documentation/blob/doc-create-NIC-passthrough-and-Kata/use-cases/images/NIC%20passthrough%20Diagram.png)
+
```
I wonder if the kata 'default' diagram should be more like the `tc` based one over at https://github.com/kata-containers/documentation/blob/master/design/architecture.md#networking - possibly to show we default to `tc` in the VM and not a virtual nic per-se any more, I think? /cc @amshinde

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+#### In this guide:
+
+* [Before you start (Restrictions, Requirements and Assumptions)](#before-you-start-restrictions-requirements-and-assumptions)
```
I think we normally just place this at the top of the document as a TOC in Kata docs. Have a look at the [kata documentation guide](https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md) if you have not already - there are probably some other conventions we use that will apply to the rest of the document.

Maybe also see something like the existing [GPU guide](https://github.com/kata-containers/documentation/blob/master/use-cases/Intel-GPU-passthrough-and-Kata.md) for an example of style and flow.

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+The method described in this guide relays on your system [supporting  IOMMU.]( https://en.wikipedia.org/wiki/Input–output_memory_management_unit#Published_specifications) before you start, make sure that your system has this attribute. If it does not, you should look for a different solution then the one presented here. 
+
+If you are uncertain about your system’s support, the first two steps in [Part 1 – The Host](# Part-1-–-The-Host) should help you figure this out. 
```
hotlink is not resolving.

jodh-intel replied:
Yes, I think you need to remove the space between the `#` and `Part`.

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+   Edit grub file ( */etc/default/grub* )
+
+```bash
+vi /etc/default/grub
````
this might want to be a `$ sudo update-grub` - is that more likely @jodh-intel ?

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+#### 2. Test that the system loads with IOMMU:
+
+```bash
+dmesg | head 
````
I wonder if we can come up with a better `grep` expression to search for, or maybe even better if there is soemthing in /proc or /sys or cpuinfo etc... we should expand/elaborate this bit.

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+Get address: 
+
+```bash
+lshw -class network -businfo -numeric | grep enp0s20f1
````
we probably need to make this less system specific as well - should be possible I think

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+> 
+> 8086 1f40
+
+#### 11. Check what the new device name is, now under vfio
```
I've just noticed you have numbered the sections in the `##...` section headers - we don't normally do that, and it makes maintaining them harder - feel free to delete the number indexing.

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+#### 12.   Create the container you wish to pass the NIC into
+
+**Note:** once running the next command you will be automatically transferred into the container’s bash shell. 
```
We probably don't need this note - hopefully folks running Kata know how to run docker, and we can't explain all details to them ;-).

grahamwhaley commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+```bash
+docker run -it --runtime=kata-runtime --name vfio_con --device /dev/vfio/39 -v /dev:/dev ubuntu bash
+```
+If you aren’t sure what just happened back there, here’s a breakdown of the command:
````
Nice. I like the idea of the breakdown!

devimc commented:
@Yuval-Ai  thanks for the PR, we appreciate it and don't worry if you can't complete the document (https://github.com/kata-containers/runtime/issues/1876#issuecomment-603105747) we can work on it later

jcvenegas commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+# How To Pass a Physical NIC Into a Kata Container
+![](https://github.com/Yuval-Ai/documentation/blob/doc-create-NIC-passthrough-and-Kata/use-cases/images/NIC%20passthrough%20Diagram.png)
```
This URL points to a personal repository, could you update/change it for the one that will be here.

grahamwhaley replied:
and you can make it a relative link - just use a ./dir/file.jpg etc. - that should work, and is then more portable etc.

fyi, I use [grip](https://github.com/joeyespo/grip) locally to preview docs I am working on before pushing - helps my turnaround time.

jcvenegas commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+#### 6. Insert the address into a variable
+
+```bash
+BDF="0000:00:14.1"
````
This could be  assigned from the `lshw` output with the help of grep and awk and other parsing help something like :
`BDF=$(shw -class network -businfo -numeric | grep ${NIC} | awk 'print $1' | cut -d@ -f2)`

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+If you are uncertain about your system’s support, the first two steps in [Part 1 – The Host](# Part-1-–-The-Host) should help you figure this out. 
+
+ In my setup, I used a server with *Intel(R) Xeon(R) CPU D-1548 @ 2.00GHz* and [support for VT-d](https://en.wikipedia.org/wiki/X86_virtualization#Intel-VT-d), Operated by Ubuntu 18.04.3 LTS. The general steps proposed is this guide should match other hardware and OS with some changes.  
```
I would make this more generic - rather than "In my setup" you could say, "This guide was tested using ...".

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+ In my setup, I used a server with *Intel(R) Xeon(R) CPU D-1548 @ 2.00GHz* and [support for VT-d](https://en.wikipedia.org/wiki/X86_virtualization#Intel-VT-d), Operated by Ubuntu 18.04.3 LTS. The general steps proposed is this guide should match other hardware and OS with some changes.  
+
+This guide’s starting point is a system with an installed and working *Docker* and *Kata Containers*.
```
I think you can remove the asterisks around these two terms.

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+If that is not the case for you:
+
+\-    [How to install Kata Containers]( https://github.com/kata-containers/documentation/tree/master/install)
```
Please remove the backslashes in these two lines (and the spaces immediately after the open bracket symbols.

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+### Part 1 – The Host
+
+**Note:** all the next commands are run as a super-user. Use `$ sudo -i` to become a SU.
```
- Nit: s/all/All.
- s/SU/super-user/ (or `root`).

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,292 @@
+
+#### 1. Configure system to load with IOMMU support.
+
+   Edit grub file ( */etc/default/grub* )
```
All files need to be put in backticks, not asterisks.

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+/TODO – Change to Ubuntu
+
+```bash
+grub2-mkconfig -o /boot/efi/EFI/centos/grub.cfg
````
All commands need to start with `$ `. But if a command needs `root`, the line should start with `$ sudo `.

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,292 @@
+grub2-mkconfig -o /boot/grub2/grub.cfg
+```
+
+ > Expected result
````
We tend not to should command output unless absolutely essential, so I would drop this. If output needs to be shown, it needs to be in backticks, not using the `>` quoting mechanism.

grahamwhaley commented:
Hi @Yuval-Ai - just checking, presume you will get back to this sometime?

Yuval-Ai commented:
Hi! Yes :)
Hopefully soon.

amshinde commented:
Hi @Yuval-Ai Thanks for the PR. Can you squash all your commits into a single commit. Will make it easier to review as well.

Yuval-Ai commented:
Hi @amshinde, squashed my commits. Can you please take a look and tell me if it's fine? 
Thanks.

Yuval-Ai commented:
Hi @grahamwhaley, @amshinde.
Hope everybody’s fine these days.

I changed things as discussed, and squashed my commits. When you’ll have the time, please update me what to do next.

Wishing you the best of health,
Y

grahamwhaley commented:
Hi @Yuval-Ai - thanks for updating! github is reporting that the `README.md` file is 'conflicted', which means it has changed since you made your changes - you will need to rebase your changes on top of the master again and re-push. That'll be the next step :-)

amshinde commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+![](./images/NIC%20passthrough%20Diagram.png)
+
+
+In this guide we walk through the process of passing a physical NIC into a KC (Kata Container). The "usual" way that a KC is wired for communication is through a series of bridges and virtual NICs as can be seen in the Image above (marked as default). 
```
We don't really use the abbreviation KC, so would avoid it here.

yuvalailer replied:
Done.

amshinde commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+![](./images/NIC%20passthrough%20Diagram.png)
+
+
+In this guide we walk through the process of passing a physical NIC into a KC (Kata Container). The "usual" way that a KC is wired for communication is through a series of bridges and virtual NICs as can be seen in the Image above (marked as default). 
```
/Image/image/
I would rewrite it as a "through a bridge and virtual NIC" and drop the `a series..`

yuvalailer replied:
Done.

amshinde commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+
+## Before you start (Restrictions, Requirements and Assumptions)  
+
+The method described in this guide relays on your system [supporting  IOMMU.]( https://en.wikipedia.org/wiki/Input–output_memory_management_unit#Published_specifications) before you start, make sure that your system has this attribute. If it does not, you should look for a different solution then the one presented here. 
```
Should this be "relies" here?
/before/Before/

yuvalailer replied:
Done.

amshinde commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+> **Note:** The [XXXX:XXXX] address you are looking for is the same one you used in the previous steps.
+
+#### Bind the passed device with a driver in the container.
+/ TODO
```
Are you planning to update these?

yuvalailer replied:
No, Please see: [603105747](https://github.com/kata-containers/runtime/issues/1876#issuecomment-603105747). Thanks

egernst commented:
This seems like a worthwhile PR - I'm wondering if it'd be helpful if we applied a commit to the submitters branch in order to address any fixes that are needed, to help get this into merge-able shape.

@amshinde  WDYT?  Is it accurate at this point, and just needs editorial adjustments?

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+
+
+
+If using UEFI: (like I do) 
```
Please remove "like I do".

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+![](./images/NIC%20passthrough%20Diagram.png)
+
+
+In this guide we walk through the process of passing a physical NIC into Kata Containers. The "usual" way that a Kata Container is wired for communication is through a bridge and virtual NIC, as can be seen in the image above (marked as default).
```
Nit: It would help if you folded long lines somewhere between 70-80 characters throughout the document. That makes diffs easier to read.

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+
+In this guide we walk through the process of passing a physical NIC into Kata Containers. The "usual" way that a Kata Container is wired for communication is through a bridge and virtual NIC, as can be seen in the image above (marked as default).
+
+For some use cases, the container demands a direct link to the physical port of the host device, for example in a situation were the container is required to route high BW traffic without having support for acceleration such as SR-IOV.
```
s/BW/bandwidth/ ?

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+
+### Part 1 – The Host
+
+**Note:** All next commands are run as a super-user. Use `$ sudo -i` to become a super-user.
```
I'd remove the 2nd sentence:

```suggestion
**Note:** All the following commands are run as the super-user.
```

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+
+
+
+#### Install *lshw* and *pciutils* (Ubuntu)
```
I would consider making the title more generic (easier to read and also won't need to be updated if we later find we need to add more tools)

> #### Install required tools for Ubuntu

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,217 @@
+#### Install *lshw* and *pciutils* (Ubuntu)
+
+```bash
+$ sudo apt install lshw pciutils
````
```suggestion
$ sudo apt install -y lshw pciutils
```

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,217 @@
+#### Create the container you wish to pass the NIC into
+
+```bash
+docker run -it --runtime=kata-runtime --name <choose a name> --device /dev/vfio/<device name from previous step> -v /dev:/dev <image> <command>
````
Please add a `$ ` to the start of all commands (here and below). Also, we tend to document running docker using `sudo`.

```suggestion
$ sudo docker run -it --runtime=kata-runtime --name <choose a name> --device /dev/vfio/<device name from previous step> -v /dev:/dev <image> <command>
```

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+- ` --runtime=kata-runtime ` - tells docker to run the container as a Kata Container (if you set kata-runtime as the default runtime, this option is redundant).
+- ` --name vfio_con ` – name the container. It can be any name you choose. 
+- ` --device /dev/vfio/39 ` – this one is the important part. With this command, you tell docker to pass the NIC to the container. 
+- `-v /dev:/dev `– here you tell docker to map the devices folder of the host to the devices folder of the container.
```
I'd consider removing most of this explanation, keeping only details of `--device` and `-v`.

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,217 @@
+Update and install pciutils (Ubuntu)
+
+```bash
+$ sudo apt update ; apt install pciutils
````
Haven't we dont this already above?

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
```diff
@@ -0,0 +1,217 @@
+
+#### Find PCI address for the interface you wish to pass to the container and place it in a variable
+
+Say I want to pass **enp0s20f1** to the container (see what interfaces you have by running `$ ip link` ). 
```
I'd consider making the adding a sentce at the start of this section regarding `ip link`:

> To list available interfaces, run:
> 
> ```sh
> $ ip list
> ```
>
> Assuming you want to pass the network interface `enp0s20f1` to the container, ...

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,217 @@
+```
+
+```
+Expected results (something like):
````
This line shouldn't be inside the ` ``` ` (I'd put it above it).

jodh-intel commented on use-cases/NIC-passthrough-and-Kata.md:
````diff
@@ -0,0 +1,217 @@
+$ sudo echo 8086 1f40 | sudo tee /sys/bus/pci/drivers/vfio-pci/new_id
+```
+
+> **Note:** The ` xxxx xxxx ` format is of the correct form and not a mistake.
````
You could make this clearer by single quoting the value maybe:

```sh
$ sudo echo '8086 1f40' | ...
```

jodh-intel reviewed (commented):
Thanks for updating @Yuval-Ai. I have a few more comments...

fidencio commented:
Folks, this PR has been stale and I'm closing it as we're walking towards the **final** 1.13.0 release of the kata-containers-1.x.
From now on, all the development will happen as part of https://github.com/kata-containers/kata-containers/ and fixes should be provided to that repo.

Thanks a lot for your contribution, we're sorry that for some reason it got stale, and we hope to see you around for new adventures in the kata-containers-2.x land.
